State handling crash between card browser / note editor / template editor

I've seen this show up via ACRA, and the google play console release robot army caught it with video this time so I could figure it out.

###### Reproduction Steps

1. Open the card browser (it's empty in their tests, and they play with searches sometimes but the search was cleared)
2. Click + to add a card, enter text in front and back, template should be Basic
3. click the button bar that says "Card 1" to open the template editor
4. Click the options menu and select add to add a Card to the template, and you can mess with the template or not, but don't save - preview or not but don't save - then click the back button and agree it's okay to lose your progress
5. The Card button bar should say "Cards: Card 1 Card 2" now, but you never saved the second template
6. Click save to add the card, then click back to go back to the browser and it should crash with a JSONException string out of bounds


###### Expected Result

If you don't save the template the unsaved work should go away, you should have only Card 1 and the browser should work fine


###### Actual Result

Crash. I can reproduce this locally.

https://play.google.com/apps/publish/?account=5338425030304417978#PreLaunchReportPlace:p=com.ichi2.anki&appid=4973711737547064258&plrtab=CRASH&plrvc=20900140

FATAL EXCEPTION: AsyncTask #4
Process: com.ichi2.anki, PID: 12740
java.lang.RuntimeException: An error occurred while executing doInBackground()
	at android.os.AsyncTask$3.done(AsyncTask.java:353)
	at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
	at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
	at java.util.concurrent.FutureTask.run(FutureTask.java:271)
	at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:245)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
	at java.lang.Thread.run(Thread.java:764)
Caused by: java.lang.RuntimeException: org.json.JSONException: Index 1 out of range [0..1)
	at com.ichi2.libanki.Card.template(Card.java:342)
	at com.ichi2.libanki.Card._getQA(Card.java:290)
	at com.ichi2.anki.CardBrowser.updateSearchItemQA(CardBrowser.java:1164)
	at com.ichi2.async.DeckTask.doInBackgroundSearchCards(DeckTask.java:818)
	at com.ichi2.async.DeckTask.doInBackground(DeckTask.java:262)
	at com.ichi2.async.DeckTask.doInBackground(DeckTask.java:69)
	at android.os.AsyncTask$2.call(AsyncTask.java:333)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	... 4 more
Caused by: org.json.JSONException: Index 1 out of range [0..1)
	at org.json.JSONArray.get(JSONArray.java:293)
	at org.json.JSONArray.getJSONObject(JSONArray.java:521)
	at com.ichi2.libanki.Card.template(Card.java:337)
	... 11 more
Caused by: java.lang.IndexOutOfBoundsException: Index: 1, Size: 1
	at java.util.ArrayList.get(ArrayList.java:437)
	at org.json.JSONArray.get(JSONArray.java:287)
	... 13 more


###### Debug info
Refer to the [support page](https://ankidroid.org/docs/help.html) if you are unsure where to get the "debug info".

###### Research
*Enter an [ x ] character to confirm the points below:*

[  ] I have read the [support page](https://ankidroid.org/docs/help.html) and am reporting a bug or enhancement request specific to AnkiDroid

[  ] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue

[  ] I have searched for similar existing issues here and on the user forum

--- a/AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidApp.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidApp.java
@@ -146,7 +146,7 @@ public class AnkiDroidApp extends Application {
      * collections being upgraded to (or after) this version must run an integrity check as it will contain fixes that
      * all collections should have.
      */
-    public static final int CHECK_DB_AT_VERSION = 40;
+    public static final int CHECK_DB_AT_VERSION = 20805000;
 
     /**
      * The latest package version number that included changes to the preferences that requires handling. All

--- a/AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java
@@ -936,12 +936,13 @@ public class DeckPicker extends NavigationDrawerActivity implements
 
             if (previous < upgradeDbVersion || previous < upgradePrefsVersion) {
                 if (previous < upgradePrefsVersion && current >= upgradePrefsVersion) {
-                    Timber.d("Upgrading preferences");
+                    Timber.i("showStartupScreensAndDialogs() running upgradePreferences()");
                     CompatHelper.removeHiddenPreferences(this.getApplicationContext());
                     upgradePreferences(previous);
                 }
                 // Integrity check loads asynchronously and then restart deckpicker when finished
                 if (previous < upgradeDbVersion && current >= upgradeDbVersion) {
+                    Timber.i("showStartupScreensAndDialogs() running integrityCheck()");
                     integrityCheck();
                 } else if (previous < upgradePrefsVersion && current >= upgradePrefsVersion) {
                     // If integrityCheck() doesn't occur, but we did update preferences we should restart DeckPicker to

--- a/AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.java
@@ -335,7 +335,7 @@ public class NoteEditor extends AnkiActivity {
         this.getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_HIDDEN);
 
         Intent intent = getIntent();
-        Timber.d("onCollectionLoaded: caller: %d", mCaller);
+        Timber.d("NoteEditor() onCollectionLoaded: caller: %d", mCaller);
 
         SharedPreferences preferences = AnkiDroidApp.getSharedPrefs(getBaseContext());
 
@@ -534,7 +534,7 @@ public class NoteEditor extends AnkiActivity {
         });
 
         if (!mAddNote && mCurrentEditedCard != null) {
-            Timber.i("NoteEditor:: Edit note activity successfully started with card id %d", mCurrentEditedCard.getId());
+            Timber.i("onCollectionLoaded() Edit note activity successfully started with card id %d", mCurrentEditedCard.getId());
         }
     }
 
@@ -1008,12 +1008,14 @@ public class NoteEditor extends AnkiActivity {
         // Pass the model ID
         try {
             intent.putExtra("modelId", getCurrentlySelectedModel().getLong("id"));
+            Timber.d("showCardTemplateEditor() for model %s", intent.getLongExtra("modelId", -1L));
         } catch (JSONException e) {
            throw new RuntimeException(e);
         }
         // Also pass the card ID if not adding new note
         if (!mAddNote) {
             intent.putExtra("noteId", mCurrentEditedCard.note().getId());
+            Timber.d("showCardTemplateEditor() with note %s", mCurrentEditedCard.note().getId());
         }
         startActivityForResultWithAnimation(intent, REQUEST_TEMPLATE_EDIT, ActivityTransitionAnimation.LEFT);
     }
@@ -1021,6 +1023,7 @@ public class NoteEditor extends AnkiActivity {
 
     @Override
     protected void onActivityResult(int requestCode, int resultCode, Intent data) {
+        Timber.d("onActivityResult() with request/result: %s/%s", requestCode, resultCode);
         super.onActivityResult(requestCode, resultCode, data);
 
         if (resultCode == DeckPicker.RESULT_DB_ERROR) {
@@ -1065,6 +1068,9 @@ public class NoteEditor extends AnkiActivity {
                 if (resultCode == RESULT_OK) {
                     mReloadRequired = true;
                 }
+                // rebuild the model post-template-edit so we get the correct number of cards
+                Timber.d("onActivityResult() template edit - reloading model");
+                mEditorNote.reloadModel();
                 updateCards(mEditorNote.model());
         }
     }
@@ -1407,10 +1413,12 @@ public class NoteEditor extends AnkiActivity {
 
     /** Update the list of card templates for current note type */
     private void updateCards(JSONObject model) {
+        Timber.d("updateCards()");
         try {
             JSONArray tmpls = model.getJSONArray("tmpls");
             String cardsList = "";
             // Build comma separated list of card names
+            Timber.d("updateCards() template count is %s", tmpls.length());
             for (int i = 0; i < tmpls.length(); i++) {
                 String name = tmpls.getJSONObject(i).optString("name");
                 // If more than one card then make currently selected card underlined

--- a/AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java
+++ b/AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java
@@ -1510,7 +1510,7 @@ public class Collection {
         if (problems.size() > 0) {
             modSchemaNoCheck();
         }
-        // TODO: report problems
+        logProblems(problems);
         return (oldSize - newSize) / 1024;
     }
 
@@ -1527,6 +1527,27 @@ public class Collection {
      * Logging
      * ***********************************************************
      */
+
+    /**
+     * Track database corruption problems - AcraLimiter should quench it if it's a torrent
+     * but we will take care to limit possible data usage by limiting count we send regardless
+     *
+     * @param integrityCheckProblems list of problems, the first 10 will be logged and sent via ACRA
+     */
+    private void logProblems(ArrayList integrityCheckProblems) {
+
+        if (integrityCheckProblems.size() > 0) {
+            StringBuffer additionalInfo = new StringBuffer();
+            for (int i = 0; ((i < 10) && (integrityCheckProblems.size() > i)); i++) {
+                additionalInfo.append(integrityCheckProblems.get(i)).append("\n");
+            }
+            AnkiDroidApp.sendExceptionReport(
+                    new Exception("Problem list (limited to first 10)"), "Collection.fixIntegrity()", additionalInfo.toString());
+            Timber.i("fixIntegrity() Problem list (limited to first 10):\n%s", additionalInfo);
+        } else {
+            Timber.i("fixIntegrity() no problems found");
+        }
+    }
 
     public void log(Object... args) {
         if (!mDebugLog) {

--- a/AnkiDroid/src/main/java/com/ichi2/libanki/Note.java
+++ b/AnkiDroid/src/main/java/com/ichi2/libanki/Note.java
@@ -31,6 +31,8 @@ import java.util.List;
 import java.util.Locale;
 import java.util.Map;
 
+import timber.log.Timber;
+
 
 public class Note implements Cloneable {
 
@@ -88,6 +90,7 @@ public class Note implements Cloneable {
 
 
     public void load() {
+        Timber.d("load()");
         Cursor cursor = null;
         try {
             cursor = mCol.getDb().getDatabase()
@@ -111,6 +114,10 @@ public class Note implements Cloneable {
                 cursor.close();
             }
         }
+    }
+
+    public void reloadModel() {
+        mModel = mCol.getModels().get(mMid);
     }
 
 

